test/framework/machine_helpers.go

```diff
@@ -21,32 +21,66 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/pkg/errors"
 
+	"github.com/pkg/errors"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+type GetMachinesByClusterInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+}
+
+// GetMachinesByCluster returns the Machine objects that matches provided options for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachinesByCluster(ctx context.Context, input GetMachinesByClusterInput, options ...client.ListOption) []clusterv1.Machine {
+	opts := append(byClusterOptions(input.ClusterName, input.Namespace), options...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, opts...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+// deploymentNameMachineOptions returns a set of ListOptions that allows to get all machine objects belonging to a machine deployment.
+func deploymentNameMachineOptions(deploymentName string) []client.ListOption {
```

Comment:
To support user-defined label selector (similar to CAPV), rather than deploymentName here, you can either:
- pass in a MachineDeployment and use the Spec.Selector value to set the client ListOption
- pass in a metav1.LabelSelector directly and use it to set the client ListOption

Reply:
@detiber Thanks for your comments.
MachineDeployment label is "cluster.x-k8s.io/cluster-name" for worker machines but control-plane machines also have this label. So, if I use MD label, I will also get control-plane machines that is not managed by that MD.

What is the best way to reach the Machines managed by an MD in this case? Getting MachineSet and then Machines?

Reply:
@sedefsavas If the selector on the MachineDeployment is only matching on the `cluster.x-k8s.io/cluster-name` label, then I think that is a bug against CAPV. This selector (if defined) should be a unique selector for Machines belonging to the MachineDeployment.

I'll follow up with an issue against CAPV for that.

Reply:
Yes, I confirmed that this is the case.

This is CAPD MachineDeployment match labels:
```
     matchLabels:
        cluster.x-k8s.io/cluster-name: cluster-kqqvii
        cluster.x-k8s.io/deployment-name: cluster-kqqvii-md-0
```
And this is for CAPV:
```
      matchLabels:
        cluster.x-k8s.io/cluster-name: cluster-lv5432
```

Reply:
I created https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/897 to track fixing the CAPV templates.